A1

Can we determine a pattern from market basket analysis where a certain item will likely cause a customer purchase the next item?

A2

One goal of this analysis is to provide the company with the appropriate information to understand their customers in order to sell the right products to their customers 

B1

Market basket analysis works by finding patterns in transactions that occur frequently. It is typically stated as 'IF {item1} THEN {item2}. The IF statement is called antecedent and the THEN statement is called the consequent. It is then stated as a probability between that occurance and all the possible transactions (Datacamp, 2022).

The expected outcome will be discovering a pattern of items that customers purchase and using that information to market the conseuqent item when the antecedent item is triggered.

B2

Upon observing the first row of the teleco_market_basket dataset, which we can refer as the transaction of items purchased by a customer:

Logitech M510 Wireless mouse	HP 63 Ink	HP 65 ink	nonda USB C to USB Adapter	10ft iPHone Charger Cable	HP 902XL ink	Creative Pebble 2.0 Speakers	Cleaning Gel Universal Dust Cleaner	Micro Center 32GB Memory card	YUNSONG 3pack 6ft Nylon Lightning Cable	TopMate C5 Laptop Cooler pad	Apple USB-C Charger cable	HyperX Cloud Stinger Headset	TONOR USB Gaming Microphone	Dust-Off Compressed Gas 2 pack	3A USB Type C Cable 3 pack 6FT	HOVAMP iPhone charger	SanDisk Ultra 128GB card	FEEL2NICE 5 pack 10ft Lighning cable	FEIYOLD Blue light Blocking Glasses

The transaction above is a list of 20 items.

B3

An assumption for Market basket analysis is the transactional dataset contains relations with each other and that the use of association rules will find the relationship between the items. 

C1

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
telco = pd.read_csv('teleco_market_basket.csv')

In [3]:
telco

Item01  \
0                                        NaN   
1               Logitech M510 Wireless mouse   
2                                        NaN   
3      Apple Lightning to Digital AV Adapter   
4                                        NaN   
...                                      ...   
14997         Falcon Dust Off Compressed Gas   
14998                                    NaN   
14999                            HP 63XL Ink   
15000                                    NaN   
15001                           Apple Pencil   

                                 Item02                 Item03  \
0                                   NaN                    NaN   
1                             HP 63 Ink              HP 65 ink   
2                                   NaN                    NaN   
3      TP-Link AC1750 Smart WiFi Router           Apple Pencil   
4                                   NaN                    NaN   
...                                 ...                    ...   
14997                               NaN                    NaN   
14998                               NaN                    NaN   
14999         Apple USB-C Charger cable                    NaN   
15000                               NaN                    NaN   
15001          SanDisk Ultra 128GB card  RUNMUS Gaming Headset   

                             Item04                     Item05        Item06  \
0                               NaN                        NaN           NaN   
1        nonda USB C to USB Adapter  10ft iPHone Charger Cable  HP 902XL ink   
2                               NaN                        NaN           NaN   
3                               NaN                        NaN           NaN   
4                               NaN                        NaN           NaN   
...                             ...                        ...           ...   
14997                           NaN                        NaN           NaN   
14998                           NaN                        NaN           NaN   
14999                           NaN                        NaN           NaN   
15000                           NaN                        NaN           NaN   
15001  TopMate C5 Laptop Cooler pad                        NaN           NaN   

                             Item07                               Item08  \
0                               NaN                                  NaN   
1      Creative Pebble 2.0 Speakers  Cleaning Gel Universal Dust Cleaner   
2                               NaN                                  NaN   
3                               NaN                                  NaN   
4                               NaN                                  NaN   
...                             ...                                  ...   
14997                           NaN                                  NaN   
14998                           NaN                                  NaN   
14999                           NaN                                  NaN   
15000                           NaN                                  NaN   
15001                           NaN                                  NaN   

                              Item09                                   Item10  \
0                                NaN                                      NaN   
1      Micro Center 32GB Memory card  YUNSONG 3pack 6ft Nylon Lightning Cable   
2                                NaN                                      NaN   
3                                NaN                                      NaN   
4                                NaN                                      NaN   
...                              ...                                      ...   
14997                            NaN                                      NaN   
14998                            NaN                                      NaN   
14999                            NaN                             

In [4]:
telco.shape

(15002, 20)

In [5]:
telco.describe()

Item01                          Item02  \
count                             7501                            5747   
unique                             115                             117   
top     Dust-Off Compressed Gas 2 pack  Dust-Off Compressed Gas 2 pack   
freq                               577                             484   

                                Item03                          Item04  \
count                             4389                            3345   
unique                             115                             114   
top     Dust-Off Compressed Gas 2 pack  Dust-Off Compressed Gas 2 pack   
freq                               375                             201   

                           Item05                 Item06  \
count                        2529                   1864   
unique                        110                    106   
top     Apple USB-C Charger cable  USB 2.0 Printer cable   
freq                          153                    107   

                           Item07                     Item08  \
count                        1369                        981   
unique                        102                         97   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                           96                         67   

                           Item09                     Item10  \
count                         654                        395   
unique                         88                         80   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                           57                         31   

                              Item11                     Item12  \
count                            256                        154   
unique                            66                         50   
top     TopMate C5 Laptop Cooler pad  Apple USB-C Charger cable   
freq                              22                         15   

                           Item13                     Item14  \
count                          87                         47   
unique                         43                         28   
top     Apple USB-C Charger cable  Apple USB-C Charger cable   
freq                            8                          4   

                                    Item15                      Item16  \
count                                   25                           8   
unique                                  19                           8   
top     ARRIS SURFboard SB8200 Cable Modem  SanDisk Extreme 128GB card   
freq                                     3                           1   

                          Item17                                      Item18  \
count                          4                                           4   
unique                         3                                           3   
top     SanDisk Ultra 128GB card  Brother Genuine High Yield Toner Cartridge   
freq                           2                                           2   

                                      Item19  \
count                                      3   
unique                                     3   
top     FEEL2NICE 5 pack 10ft Lighning cable   
freq                                       1   

                                     Item20  
count                                     1  
unique                                    1  
top     FEIYOLD Blue light Blocking Glasses  
freq                                      1

In [6]:
telco.isnull().sum()

Item01     7501
Item02     9255
Item03    10613
Item04    11657
Item05    12473
Item06    13138
Item07    13633
Item08    14021
Item09    14348
Item10    14607
Item11    14746
Item12    14848
Item13    14915
Item14    14955
Item15    14977
Item16    14994
Item17    14998
Item18    14998
Item19    14999
Item20    15001
dtype: int64

In [7]:
telco.dropna(how='all', inplace=True)

In [8]:
telco

Item01  \
1                   Logitech M510 Wireless mouse   
3          Apple Lightning to Digital AV Adapter   
5      UNEN Mfi Certified 5-pack Lightning Cable   
7                            Cat8 Ethernet Cable   
9                 Dust-Off Compressed Gas 2 pack   
...                                          ...   
14993               SanDisk 32GB Ultra SDHC card   
14995      Apple Lightning to Digital AV Adapter   
14997             Falcon Dust Off Compressed Gas   
14999                                HP 63XL Ink   
15001                               Apple Pencil   

                                     Item02                        Item03  \
1                                 HP 63 Ink                     HP 65 ink   
3          TP-Link AC1750 Smart WiFi Router                  Apple Pencil   
5                                       NaN                           NaN   
7                                 HP 65 ink                           NaN   
9             Screen Mom Screen Cleaner kit    Moread HDMI to VGA Adapter   
...                                     ...                           ...   
14993                 Vsco 70 pack stickers  SanDisk 128GB microSDXC card   
14995  Nylon Braided Lightning to USB cable                  Apple Pencil   
14997                                   NaN                           NaN   
14999             Apple USB-C Charger cable                           NaN   
15001              SanDisk Ultra 128GB card         RUNMUS Gaming Headset   

                             Item04                              Item05  \
1        nonda USB C to USB Adapter           10ft iPHone Charger Cable   
3                               NaN                                 NaN   
5                               NaN                                 NaN   
7                               NaN                                 NaN   
9             HP 62XL Tri-Color ink           Apple USB-C Charger cable   
...                             ...                                 ...   
14993                           NaN                                 NaN   
14995         USB 2.0 Printer cable  ARRIS SURFboard SB8200 Cable Modem   
14997                           NaN                                 NaN   
14999                           NaN                                 NaN   
15001  TopMate C5 Laptop Cooler pad                                 NaN   

                          Item06                        Item07  \
1                   HP 902XL ink  Creative Pebble 2.0 Speakers   
3                            NaN                           NaN   
5                            NaN                           NaN   
7                            NaN                           NaN   
9                            NaN                           NaN   
...                          ...                           ...   
14993                        NaN                           NaN   
14995  Apple USB-C Charger cable                           NaN   
14997                        NaN                           NaN   
14999                        NaN                           NaN   
15001                        NaN                           NaN   

                                    Item08                         Item09  \
1      Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
3                                      NaN                            NaN   
5                                      NaN                            NaN   
7                                      NaN                            NaN   
9                                      NaN                            NaN   
...                                    ...                            ...   
14993                                  NaN                            NaN   
14995                                  NaN                            NaN   
14997                                  NaN                            NaN   
14999                       

telco.fillna(0, inplace=True)

In [9]:
telco.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7501 entries, 1 to 15001
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Item01  7501 non-null   object
 1   Item02  5747 non-null   object
 2   Item03  4389 non-null   object
 3   Item04  3345 non-null   object
 4   Item05  2529 non-null   object
 5   Item06  1864 non-null   object
 6   Item07  1369 non-null   object
 7   Item08  981 non-null    object
 8   Item09  654 non-null    object
 9   Item10  395 non-null    object
 10  Item11  256 non-null    object
 11  Item12  154 non-null    object
 12  Item13  87 non-null     object
 13  Item14  47 non-null     object
 14  Item15  25 non-null     object
 15  Item16  8 non-null      object
 16  Item17  4 non-null      object
 17  Item18  4 non-null      object
 18  Item19  3 non-null      object
 19  Item20  1 non-null      object
dtypes: object(20)
memory usage: 1.2+ MB


In [10]:
telco.dropna(subset=['Item01'], inplace=True)
telco_list = telco.T.apply(lambda x: x.dropna().tolist()).tolist()

telco_list

In [11]:
telco_cleaned = pd.DataFrame(telco_list)

In [12]:
telco_cleaned

0   \
0                  Logitech M510 Wireless mouse   
1         Apple Lightning to Digital AV Adapter   
2     UNEN Mfi Certified 5-pack Lightning Cable   
3                           Cat8 Ethernet Cable   
4                Dust-Off Compressed Gas 2 pack   
...                                         ...   
7496               SanDisk 32GB Ultra SDHC card   
7497      Apple Lightning to Digital AV Adapter   
7498             Falcon Dust Off Compressed Gas   
7499                                HP 63XL Ink   
7500                               Apple Pencil   

                                        1                             2   \
0                                HP 63 Ink                     HP 65 ink   
1         TP-Link AC1750 Smart WiFi Router                  Apple Pencil   
2                                     None                          None   
3                                HP 65 ink                          None   
4            Screen Mom Screen Cleaner kit    Moread HDMI to VGA Adapter   
...                                    ...                           ...   
7496                 Vsco 70 pack stickers  SanDisk 128GB microSDXC card   
7497  Nylon Braided Lightning to USB cable                  Apple Pencil   
7498                                  None                          None   
7499             Apple USB-C Charger cable                          None   
7500              SanDisk Ultra 128GB card         RUNMUS Gaming Headset   

                                3                                   4   \
0       nonda USB C to USB Adapter           10ft iPHone Charger Cable   
1                             None                                None   
2                             None                                None   
3                             None                                None   
4            HP 62XL Tri-Color ink           Apple USB-C Charger cable   
...                            ...                                 ...   
7496                          None                                None   
7497         USB 2.0 Printer cable  ARRIS SURFboard SB8200 Cable Modem   
7498                          None                                None   
7499                          None                                None   
7500  TopMate C5 Laptop Cooler pad                                None   

                             5                             6   \
0                  HP 902XL ink  Creative Pebble 2.0 Speakers   
1                          None                          None   
2                          None                          None   
3                          None                          None   
4                          None                          None   
...                         ...                           ...   
7496                       None                          None   
7497  Apple USB-C Charger cable                          None   
7498                       None                          None   
7499                       None                          None   
7500                       None                          None   

                                       7                              8   \
0     Cleaning Gel Universal Dust Cleaner  Micro Center 32GB Memory card   
1                                    None                           None   
2                                    None                           None   
3                                    None                           None   
4                                    None                           None   
...                                   ...                            ...   
7496                                 None                           None   
7497                                 None                           None   
7498                                 None                           None   
7499                                 None                           None   
7500         

In [13]:
telco_cleaned.columns

RangeIndex(start=0, stop=20, step=1)

In [14]:
telco_cleaned.to_csv('telco_cleaned')

C2

In [15]:
!pip install mlxtend  

In [16]:
from mlxtend.frequent_patterns import apriori, association_rules
from mlxtend.preprocessing import TransactionEncoder

In [17]:
encoder = TransactionEncoder()
onehot = encoder.fit(telco_list).transform(telco_list)
onehot = pd.DataFrame(onehot, columns=encoder.columns_)

In [18]:
onehot.to_csv('onehot.csv')

In [19]:
frequent_itemsets = apriori(onehot, use_colnames=True, min_support=0.001, max_len =3)

In [20]:
rules = association_rules(frequent_itemsets)
rules

antecedents  \
0   (10ft iPHone Charger Cable 2 Pack, VicTsing Wi...   
1   (iPhone 11 case, 5pack Nylon Braided USB C cab...   
2   (Cat8 Ethernet Cable, Anker 2-in-1 USB Card Re...   
3   (TP-Link AC1750 Smart WiFi Router, Anker 2-in-...   
4   (Brother Genuine High Yield Toner Cartridge, F...   
5   (SanDisk Extreme 256GB card, Cat8 Ethernet Cable)   
6   (SanDisk Ultra 128GB card, FEEL2NICE 5 pack 10...   
7   (Premium Nylon USB Cable, SanDisk Ultra 400GB ...   
8   (SanDisk Extreme 256GB card, SanDisk Ultra 64G...   
9     (iPhone 11 case, Screen Mom Screen Cleaner kit)   
10     (SAMSUNG 128GB card, SanDisk Ultra 128GB card)   
11          (SanDisk 128GB card, Stylus Pen for iPad)   

                           consequents  antecedent support  \
0     (Dust-Off Compressed Gas 2 pack)            0.002000   
1                        (HP 63XL Ink)            0.002666   
2     (Dust-Off Compressed Gas 2 pack)            0.001733   
3      (Screen Mom Screen Cleaner kit)            0.001600   
4   (VIVO Dual LCD Monitor Desk mount)            0.001466   
5     (Dust-Off Compressed Gas 2 pack)            0.001466   
6     (Dust-Off Compressed Gas 2 pack)            0.001200   
7     (Dust-Off Compressed Gas 2 pack)            0.001466   
8     (Dust-Off Compressed Gas 2 pack)            0.001866   
9       (Logitech M510 Wireless mouse)            0.001866   
10     (Screen Mom Screen Cleaner kit)            0.001466   
11  (VIVO Dual LCD Monitor Desk mount)            0.002133   

    consequent support   support  confidence       lift  leverage  conviction  
0             0.238368  0.001866    0.933333   3.915511  0.001390   11.424477  
1             0.079323  0.002533    0.950000  11.976387  0.002321   18.413545  
2             0.238368  0.001466    0.846154   3.549776  0.001053    4.950607  
3             0.129583  0.001333    0.833333   6.430898  0.001126    5.222504  
4             0.174110  0.001200    0.818182   4.699220  0.000945    4.542394  
5             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
6             0.238368  0.001067    0.888889   3.729058  0.000781    6.854686  
7             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
8             0.238368  0.001600    0.857143   3.595877  0.001155    5.331422  
9             0.071457  0.001600    0.857143  11.995203  0.001466    6.499800  
10            0.129583  0.001200    0.818182   6.313973  0.001010    4.787295  
11            0.174110  0.001733    0.812500   4.666587  0.001362    4.404746

C3

In [21]:
values_1 = (rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']])
values_1

antecedents  \
0   (10ft iPHone Charger Cable 2 Pack, VicTsing Wi...   
1   (iPhone 11 case, 5pack Nylon Braided USB C cab...   
2   (Cat8 Ethernet Cable, Anker 2-in-1 USB Card Re...   
3   (TP-Link AC1750 Smart WiFi Router, Anker 2-in-...   
4   (Brother Genuine High Yield Toner Cartridge, F...   
5   (SanDisk Extreme 256GB card, Cat8 Ethernet Cable)   
6   (SanDisk Ultra 128GB card, FEEL2NICE 5 pack 10...   
7   (Premium Nylon USB Cable, SanDisk Ultra 400GB ...   
8   (SanDisk Extreme 256GB card, SanDisk Ultra 64G...   
9     (iPhone 11 case, Screen Mom Screen Cleaner kit)   
10     (SAMSUNG 128GB card, SanDisk Ultra 128GB card)   
11          (SanDisk 128GB card, Stylus Pen for iPad)   

                           consequents   support  confidence       lift  
0     (Dust-Off Compressed Gas 2 pack)  0.001866    0.933333   3.915511  
1                        (HP 63XL Ink)  0.002533    0.950000  11.976387  
2     (Dust-Off Compressed Gas 2 pack)  0.001466    0.846154   3.549776  
3      (Screen Mom Screen Cleaner kit)  0.001333    0.833333   6.430898  
4   (VIVO Dual LCD Monitor Desk mount)  0.001200    0.818182   4.699220  
5     (Dust-Off Compressed Gas 2 pack)  0.001200    0.818182   3.432428  
6     (Dust-Off Compressed Gas 2 pack)  0.001067    0.888889   3.729058  
7     (Dust-Off Compressed Gas 2 pack)  0.001200    0.818182   3.432428  
8     (Dust-Off Compressed Gas 2 pack)  0.001600    0.857143   3.595877  
9       (Logitech M510 Wireless mouse)  0.001600    0.857143  11.995203  
10     (Screen Mom Screen Cleaner kit)  0.001200    0.818182   6.313973  
11  (VIVO Dual LCD Monitor Desk mount)  0.001733    0.812500   4.666587

C4

In [22]:
rules.sort_values('support', ascending=False)

antecedents  \
1   (iPhone 11 case, 5pack Nylon Braided USB C cab...   
0   (10ft iPHone Charger Cable 2 Pack, VicTsing Wi...   
11          (SanDisk 128GB card, Stylus Pen for iPad)   
8   (SanDisk Extreme 256GB card, SanDisk Ultra 64G...   
9     (iPhone 11 case, Screen Mom Screen Cleaner kit)   
2   (Cat8 Ethernet Cable, Anker 2-in-1 USB Card Re...   
3   (TP-Link AC1750 Smart WiFi Router, Anker 2-in-...   
4   (Brother Genuine High Yield Toner Cartridge, F...   
5   (SanDisk Extreme 256GB card, Cat8 Ethernet Cable)   
7   (Premium Nylon USB Cable, SanDisk Ultra 400GB ...   
10     (SAMSUNG 128GB card, SanDisk Ultra 128GB card)   
6   (SanDisk Ultra 128GB card, FEEL2NICE 5 pack 10...   

                           consequents  antecedent support  \
1                        (HP 63XL Ink)            0.002666   
0     (Dust-Off Compressed Gas 2 pack)            0.002000   
11  (VIVO Dual LCD Monitor Desk mount)            0.002133   
8     (Dust-Off Compressed Gas 2 pack)            0.001866   
9       (Logitech M510 Wireless mouse)            0.001866   
2     (Dust-Off Compressed Gas 2 pack)            0.001733   
3      (Screen Mom Screen Cleaner kit)            0.001600   
4   (VIVO Dual LCD Monitor Desk mount)            0.001466   
5     (Dust-Off Compressed Gas 2 pack)            0.001466   
7     (Dust-Off Compressed Gas 2 pack)            0.001466   
10     (Screen Mom Screen Cleaner kit)            0.001466   
6     (Dust-Off Compressed Gas 2 pack)            0.001200   

    consequent support   support  confidence       lift  leverage  conviction  
1             0.079323  0.002533    0.950000  11.976387  0.002321   18.413545  
0             0.238368  0.001866    0.933333   3.915511  0.001390   11.424477  
11            0.174110  0.001733    0.812500   4.666587  0.001362    4.404746  
8             0.238368  0.001600    0.857143   3.595877  0.001155    5.331422  
9             0.071457  0.001600    0.857143  11.995203  0.001466    6.499800  
2             0.238368  0.001466    0.846154   3.549776  0.001053    4.950607  
3             0.129583  0.001333    0.833333   6.430898  0.001126    5.222504  
4             0.174110  0.001200    0.818182   4.699220  0.000945    4.542394  
5             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
7             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
10            0.129583  0.001200    0.818182   6.313973  0.001010    4.787295  
6             0.238368  0.001067    0.888889   3.729058  0.000781    6.854686

In [23]:
rules.sort_values('confidence', ascending=False)

antecedents  \
1   (iPhone 11 case, 5pack Nylon Braided USB C cab...   
0   (10ft iPHone Charger Cable 2 Pack, VicTsing Wi...   
6   (SanDisk Ultra 128GB card, FEEL2NICE 5 pack 10...   
8   (SanDisk Extreme 256GB card, SanDisk Ultra 64G...   
9     (iPhone 11 case, Screen Mom Screen Cleaner kit)   
2   (Cat8 Ethernet Cable, Anker 2-in-1 USB Card Re...   
3   (TP-Link AC1750 Smart WiFi Router, Anker 2-in-...   
4   (Brother Genuine High Yield Toner Cartridge, F...   
5   (SanDisk Extreme 256GB card, Cat8 Ethernet Cable)   
7   (Premium Nylon USB Cable, SanDisk Ultra 400GB ...   
10     (SAMSUNG 128GB card, SanDisk Ultra 128GB card)   
11          (SanDisk 128GB card, Stylus Pen for iPad)   

                           consequents  antecedent support  \
1                        (HP 63XL Ink)            0.002666   
0     (Dust-Off Compressed Gas 2 pack)            0.002000   
6     (Dust-Off Compressed Gas 2 pack)            0.001200   
8     (Dust-Off Compressed Gas 2 pack)            0.001866   
9       (Logitech M510 Wireless mouse)            0.001866   
2     (Dust-Off Compressed Gas 2 pack)            0.001733   
3      (Screen Mom Screen Cleaner kit)            0.001600   
4   (VIVO Dual LCD Monitor Desk mount)            0.001466   
5     (Dust-Off Compressed Gas 2 pack)            0.001466   
7     (Dust-Off Compressed Gas 2 pack)            0.001466   
10     (Screen Mom Screen Cleaner kit)            0.001466   
11  (VIVO Dual LCD Monitor Desk mount)            0.002133   

    consequent support   support  confidence       lift  leverage  conviction  
1             0.079323  0.002533    0.950000  11.976387  0.002321   18.413545  
0             0.238368  0.001866    0.933333   3.915511  0.001390   11.424477  
6             0.238368  0.001067    0.888889   3.729058  0.000781    6.854686  
8             0.238368  0.001600    0.857143   3.595877  0.001155    5.331422  
9             0.071457  0.001600    0.857143  11.995203  0.001466    6.499800  
2             0.238368  0.001466    0.846154   3.549776  0.001053    4.950607  
3             0.129583  0.001333    0.833333   6.430898  0.001126    5.222504  
4             0.174110  0.001200    0.818182   4.699220  0.000945    4.542394  
5             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
7             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
10            0.129583  0.001200    0.818182   6.313973  0.001010    4.787295  
11            0.174110  0.001733    0.812500   4.666587  0.001362    4.404746

In [24]:
rules.sort_values('lift', ascending=False)

antecedents  \
9     (iPhone 11 case, Screen Mom Screen Cleaner kit)   
1   (iPhone 11 case, 5pack Nylon Braided USB C cab...   
3   (TP-Link AC1750 Smart WiFi Router, Anker 2-in-...   
10     (SAMSUNG 128GB card, SanDisk Ultra 128GB card)   
4   (Brother Genuine High Yield Toner Cartridge, F...   
11          (SanDisk 128GB card, Stylus Pen for iPad)   
0   (10ft iPHone Charger Cable 2 Pack, VicTsing Wi...   
6   (SanDisk Ultra 128GB card, FEEL2NICE 5 pack 10...   
8   (SanDisk Extreme 256GB card, SanDisk Ultra 64G...   
2   (Cat8 Ethernet Cable, Anker 2-in-1 USB Card Re...   
5   (SanDisk Extreme 256GB card, Cat8 Ethernet Cable)   
7   (Premium Nylon USB Cable, SanDisk Ultra 400GB ...   

                           consequents  antecedent support  \
9       (Logitech M510 Wireless mouse)            0.001866   
1                        (HP 63XL Ink)            0.002666   
3      (Screen Mom Screen Cleaner kit)            0.001600   
10     (Screen Mom Screen Cleaner kit)            0.001466   
4   (VIVO Dual LCD Monitor Desk mount)            0.001466   
11  (VIVO Dual LCD Monitor Desk mount)            0.002133   
0     (Dust-Off Compressed Gas 2 pack)            0.002000   
6     (Dust-Off Compressed Gas 2 pack)            0.001200   
8     (Dust-Off Compressed Gas 2 pack)            0.001866   
2     (Dust-Off Compressed Gas 2 pack)            0.001733   
5     (Dust-Off Compressed Gas 2 pack)            0.001466   
7     (Dust-Off Compressed Gas 2 pack)            0.001466   

    consequent support   support  confidence       lift  leverage  conviction  
9             0.071457  0.001600    0.857143  11.995203  0.001466    6.499800  
1             0.079323  0.002533    0.950000  11.976387  0.002321   18.413545  
3             0.129583  0.001333    0.833333   6.430898  0.001126    5.222504  
10            0.129583  0.001200    0.818182   6.313973  0.001010    4.787295  
4             0.174110  0.001200    0.818182   4.699220  0.000945    4.542394  
11            0.174110  0.001733    0.812500   4.666587  0.001362    4.404746  
0             0.238368  0.001866    0.933333   3.915511  0.001390   11.424477  
6             0.238368  0.001067    0.888889   3.729058  0.000781    6.854686  
8             0.238368  0.001600    0.857143   3.595877  0.001155    5.331422  
2             0.238368  0.001466    0.846154   3.549776  0.001053    4.950607  
5             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975  
7             0.238368  0.001200    0.818182   3.432428  0.000850    4.188975

Top 3 rules generated:

1) If '5pack Nylon Braided USB C cables,iPhone 11 case' Then 'HP 63XL Ink'
    support = 0.002533
    confidence = 0.950000
    lift = 11.97

2) If 'VicTsing Wireless mouse, 10ft iPHone Charger Cable 2 Pack' Then 'Dust-Off Compressed Gas 2 pack'
    support = 0.001866
    confidence = 0.93333
    lift = 3.91

3) If 'Stylus Pen for iPad,SanDisk 128GB card' Then 'VIVO Dual LCD Monitor Desk mount'
    support = 0.001733
    confidence = 0.812500
    lift = 4.66

D1

Support:
The support from the dataset tells us the probabilty of an item occuring in the transaction, for an example, '5pack Nylon Braided USB C cables,iPhone 11 case' has a support = 0.002533 chance of having these items in the transaction 

Confidence:
The confidence explains the probabilty of the consequent given the antecedent. For an example there is a 0.95 confidence that given '5pack Nylon Braided USB C cables,iPhone 11 case' then a customer will also buy Then 'HP 63XL Ink'

Lift:
The lift is a ratio between Confidence of A and Support B, a lift greater than 1 explains that with item A, the chances of item B is higher. So if '5pack Nylon Braided USB C cables,iPhone 11 case ' then 'HP 63XL Ink' has a lift of 11.97, then 'HP 63XL Ink' is about 12 times as likely to be purchased. 


D2

The practical significance does prove that given an antecedent item that a customer will likely purchase the consequent based on the numbers of the confidence and lift. Although the low support is not preferred, it represents the frequency of the itemset occurring and we can still find associations between the items (McColl, 2018).

D3

Given the confidence and the lift values, we can advise the company to use this analysis to continue their campaign of promoting discounts to certain items that can help reduce customer churning. Our confidence and lift values represents the consequent items. So we advise the company to focus on discounting items with high confidence above .8 and lift greater than 1. 

E

F

Datacamp. (n.d). Introduction to Market Basket Analysis.

Roshan, B. (2020, December 22). 🍎Market Basket Analysis🍞- association rule mining with visualizations. Medium. Retrieved January 29, 2022, from https://medium.com/analytics-vidhya/market-basket-analysis-association-rule-mining-with-visualizations-cda24d537019 

G

Datacamp. (n.d). Introduction to Market Basket Analysis.

McColl, L. (2018, May 31). Market basket analysis: Understanding customer behaviour. Select Statistical Consultants. Retrieved January 29, 2022, from https://select-statistics.co.uk/blog/market-basket-analysis-understanding-customer-behaviour/ 

H

In [25]:
!jupyter --version

jupyter core     : 4.6.3
jupyter-notebook : 6.1.4
qtconsole        : 4.7.7
ipython          : 7.19.0
ipykernel        : 5.3.4
jupyter client   : 6.1.7
jupyter lab      : 2.2.6
nbconvert        : 6.0.7
ipywidgets       : 7.5.1
nbformat         : 5.0.8
traitlets        : 5.0.5
